Fix false positives in redundant_discardable_let rule

kaseken opened the pull request:
Fixes #6063

I’d appreciate any feedback or suggestions.  Thanks in advance for your time and review!

#### Changes

- Fix false positive for `@ViewBuilder` functions when ignore_swiftui_view_bodies is set to true.
- Fix false positive for #Preview when ignore_swiftui_view_bodies is set to true.
- Fix false positive for PreviewProvider when ignore_swiftui_view_bodies is set to true.

kaseken commented:
buildkite/swiftlint seems to be failing due to the following issue:
https://github.com/realm/SwiftLint/issues/6076

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift:
```diff
@@ -100,4 +147,56 @@ private extension AccessorBlockSyntax {
+            return false
+        }
+
+        if let type = binding.typeAnnotation?.type.as(SomeOrAnyTypeSyntax.self) {
```
Maybe is a `guard` as well. Then you can even combine all guards into one.

kaseken replied:
Combined into a single guard here: https://github.com/realm/SwiftLint/pull/6075/commits/b8e67bff3a616dfc8a9df6d65155074a66a4a2fe

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift:
```diff
@@ -100,4 +147,56 @@ private extension AccessorBlockSyntax {
+}
+
+private extension FunctionDeclSyntax {
+    var isViewBuilderFunction: Bool {
```
Could be inlined into `isViewBuilderFunctionBody`.

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift:
```diff
@@ -100,4 +147,56 @@ private extension AccessorBlockSyntax {
+            return false
+        }
+
+        return returnType.someOrAnySpecifier.text == "some" &&
+            returnType.constraint.as(IdentifierTypeSyntax.self)?.name.text == "View"
```
This could be a computed property on `SomeOrAnyTypeSyntax` to avoid the duplication.

kaseken replied:
Fixed here: https://github.com/realm/SwiftLint/pull/6075/commits/5e364e7cf1446aad3f38e47acad47989bab42762

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift:
```diff
@@ -100,4 +147,56 @@ private extension AccessorBlockSyntax {
+
+private extension ClosureExprSyntax {
+    var isPreviewMacroBody: Bool {
+        if let macroExpansionExpr = parent?.as(MacroExpansionExprSyntax.self),
+           macroExpansionExpr.macroName.text == "Preview" {
+            return true
+        }
+        return false
```
```suggestion
        parent?.as(MacroExpansionExprSyntax.self)?.macroName.text == "Preview"
```

SimplyDanny commented on CHANGELOG.md:
```diff
@@ -18,7 +18,8 @@
 ### Bug Fixes
 
-* None.
+* Fix false positives in `redundant_discardable_let` when `ignore_swiftui_view_bodies` is enabled.
```
Please append two spaces to the end of the description. They are required to enforce a line break in the rendered Markdown:
```suggestion
* Fix false positives of `redundant_discardable_let` rule in `@ViewBuilder` functions,
  `#Preview` macro bodies and preview provides when `ignore_swiftui_view_bodies` is
  enabled.  
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift:
```diff
@@ -32,6 +51,25 @@ struct RedundantDiscardableLetRule: Rule {
+                """),
         ],
         corrections: [
             Example("↓let _ = foo()"): Example("_ = foo()"),
```
Let's also add a correction test example.

kaseken replied:
Fixed here: https://github.com/realm/SwiftLint/pull/6075/commits/699a6582558b9f56b5ee8b4797e9dcd9a7e85ed2

SimplyDanny reviewed (changes requested):
Thank you for the contribution! The implementation looks very good already. 👍 

My comments are mainly about style and consistency. Please have a look.

kaseken commented on Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift:
```diff
@@ -100,4 +147,56 @@ private extension AccessorBlockSyntax {
+        }
+
+        guard variableDecl.modifiers.contains(keyword: .static) &&
+                variableDecl.bindingSpecifier.tokenKind == .keyword(.var) else {
```
Fixed here: https://github.com/realm/SwiftLint/pull/6075/commits/b8e67bff3a616dfc8a9df6d65155074a66a4a2fe

kaseken commented:
Thank you for your thorough review, @SimplyDanny

I've incorporated all of your suggestions.
Additionally, I’ve added more nonTriggeringExamples as unit tests to further ensure that the added isPreviewProviderBody isn’t causing any issues.

I’d appreciate it if you could take another look when you have time. Thanks in advance!

SimplyDanny commented on CHANGELOG.md:
```diff
@@ -24,6 +24,10 @@
+* Fix false positives of `redundant_discardable_let` rule in `@ViewBuilder` functions,
+  `#Preview` macro bodies and preview provides when `ignore_swiftui_view_bodies` is
+  enabled.  
+  [kaseken](https://github.com/kaseken)
```
Please add a link to the issue as well.

kaseken replied:
Added issue link here: https://github.com/realm/SwiftLint/pull/6075/commits/3da5c81f8282731a88fe33c5349c8fa8939fd006

kaseken replied:
Added issue link.
https://github.com/realm/SwiftLint/pull/6075/commits/103a1c079bf45046ef3ce5872ed63cb58532099a

SimplyDanny commented on CHANGELOG.md:
```diff
@@ -24,6 +24,10 @@
   or other error.  
   [Martin Redington](https://github.com/mildm8nnered)
   [#6052](https://github.com/realm/SwiftLint/issues/6052)
+* Fix false positives of `redundant_discardable_let` rule in `@ViewBuilder` functions,
```
Needs an empty line above.

Also I think this should go into the "Enhancements" section.

kaseken replied:
Added empty line and moved to enhancements section here.
https://github.com/realm/SwiftLint/pull/6075/commits/3da5c81f8282731a88fe33c5349c8fa8939fd006

SimplyDanny commented on CHANGELOG.md:
```diff
@@ -24,6 +24,10 @@
   [Martin Redington](https://github.com/mildm8nnered)
   [#6052](https://github.com/realm/SwiftLint/issues/6052)
+* Fix false positives of `redundant_discardable_let` rule in `@ViewBuilder` functions,
+  `#Preview` macro bodies and preview provides when `ignore_swiftui_view_bodies` is
```
```suggestion
  `#Preview` macro bodies and preview providers when `ignore_swiftui_view_bodies` is
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift:
```diff
@@ -32,10 +51,87 @@ struct RedundantDiscardableLetRule: Rule {
+                    ↓let _ = foo()
+                    Text("Hello, World!")
+                }
+                """, configuration: ["ignore_swiftui_view_bodies": true], excludeFromDocumentation: true),
```
👍

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Style/RedundantDiscardableLetRule.swift:
```diff
@@ -32,10 +51,87 @@ struct RedundantDiscardableLetRule: Rule {
+                        _ = foo()
+                        Text("Hello, World!")
+                    }
+                    """),
```
I think that two examples are enough to show that corrections work. Please consider adding another example with the option enabled to show that no correction would be applied.

SimplyDanny reviewed (commented):
Thanks for the updates. There are a few more points to consider.

kaseken commented:
@SimplyDanny Thanks for reviewing!  I've fixed all the issues.

Some CI jobs ('buildkite/swiftlint' and ''buildkite/swiftlint/danger') seems to be failing due to an error unrelated to this PR.

```sh
bundle exec danger --verbose
--
  | To use retry middleware with Faraday v2.0+, install `faraday-retry` gem
  | bundler: failed to load command: danger (/Users/administrator/.rbenv/versions/3.4.2/bin/danger)
  | /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/octokit-5.6.1/lib/octokit/response/raise_error.rb:14:in 'Octokit::Response::RaiseError#on_complete': GET https://api.github.com/repos/realm/SwiftLint/pulls/6075: 401 - Bad credentials // See: https://docs.github.com/rest (Octokit::Unauthorized)
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/faraday-2.7.4/lib/faraday/middleware.rb:18:in 'block in Faraday::Middleware#call'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/faraday-2.7.4/lib/faraday/response.rb:42:in 'Faraday::Response#on_complete'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/faraday-2.7.4/lib/faraday/middleware.rb:17:in 'Faraday::Middleware#call'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/octokit-5.6.1/lib/octokit/middleware/follow_redirects.rb:73:in 'Octokit::Middleware::FollowRedirects#perform_with_redirection'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/octokit-5.6.1/lib/octokit/middleware/follow_redirects.rb:61:in 'Octokit::Middleware::FollowRedirects#call'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/faraday-2.7.4/lib/faraday/rack_builder.rb:153:in 'Faraday::RackBuilder#build_response'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/faraday-2.7.4/lib/faraday/connection.rb:444:in 'Faraday::Connection#run_request'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/faraday-2.7.4/lib/faraday/connection.rb:200:in 'Faraday::Connection#get'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/sawyer-0.9.2/lib/sawyer/agent.rb:99:in 'Sawyer::Agent#call'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/octokit-5.6.1/lib/octokit/connection.rb:156:in 'Octokit::Connection#request'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/octokit-5.6.1/lib/octokit/connection.rb:19:in 'Octokit::Connection#get'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/octokit-5.6.1/lib/octokit/client/pull_requests.rb:33:in 'Octokit::Client::PullRequests#pull_request'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/danger-9.2.0/lib/danger/request_sources/github/github.rb:117:in 'Danger::RequestSources::GitHub#fetch_details'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/danger-9.2.0/lib/danger/danger_core/environment_manager.rb:53:in 'Danger::EnvironmentManager#fill_environment_vars'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/danger-9.2.0/lib/danger/danger_core/dangerfile.rb:280:in 'Danger::Dangerfile#run'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/danger-9.2.0/lib/danger/danger_core/executor.rb:28:in 'Danger::Executor#run'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/danger-9.2.0/lib/danger/commands/runner.rb:73:in 'Danger::Runner#run'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/claide-1.1.0/lib/claide/command.rb:334:in 'CLAide::Command.run'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/danger-9.2.0/bin/danger:5:in '<top (required)>'
  | from /Users/administrator/.rbenv/versions/3.4.2/bin/danger:25:in 'Kernel#load'
  | from /Users/administrator/.rbenv/versions/3.4.2/bin/danger:25:in '<top (required)>'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/bundler-2.4.12/lib/bundler/cli/exec.rb:58:in 'Kernel.load'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/bundler-2.4.12/lib/bundler/cli/exec.rb:58:in 'Bundler::CLI::Exec#kernel_load'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/bundler-2.4.12/lib/bundler/cli/exec.rb:23:in 'Bundler::CLI::Exec#run'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/bundler-2.4.12/lib/bundler/cli.rb:492:in 'Bundler::CLI#exec'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/bundler-2.4.12/lib/bundler/vendor/thor/lib/thor/command.rb:27:in 'Bundler::Thor::Command#run'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/bundler-2.4.12/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in 'Bundler::Thor::Invocation#invoke_command'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/bundler-2.4.12/lib/bundler/vendor/thor/lib/thor.rb:392:in 'Bundler::Thor.dispatch'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/bundler-2.4.12/lib/bundler/cli.rb:34:in 'Bundler::CLI.dispatch'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/bundler-2.4.12/lib/bundler/vendor/thor/lib/thor/base.rb:485:in 'Bundler::Thor::Base::ClassMethods#start'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/bundler-2.4.12/lib/bundler/cli.rb:28:in 'Bundler::CLI.start'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/bundler-2.4.12/exe/bundle:45:in 'block in <top (required)>'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/bundler-2.4.12/lib/bundler/friendly_errors.rb:117:in 'Bundler.with_friendly_errors'
  | from /Users/administrator/.rbenv/versions/3.4.2/lib/ruby/gems/3.4.0/gems/bundler-2.4.12/exe/bundle:33:in '<top (required)>'
  | from /Users/administrator/.rbenv/versions/3.4.2/bin/bundle:25:in 'Kernel#load'
  | from /Users/administrator/.rbenv/versions/3.4.2/bin/bundle:25:in '<main>'
  | make: *** [oss_scan] Error 1
  | 🚨 Error: The command exited with status 2
  | user command error: exit status 2
```
